Assets/BossRoom/Scripts/Shared/Infrastructure/DIScope.cs

```diff
@@ -0,0 +1,333 @@
+                m_DisposableGroup.Dispose();
+                m_Disposed = true;
+            }
+        }
+
+        public T Resolve<T>()
+            where T : class
+        {
+            if (!m_ScopeConstructionComplete)
+                throw new ScopeNotFinalizedException(
+                    $"Trying to Resolve type {typeof(T)}, but the DISCope is not yet finalized! You should call FinalizeScopeConstruction before any of the Resolve calls.");
+            //if we have this type as lazy-bound instance - we are going to instantiate it now
+            if (m_LazyBindDescriptors.TryGetValue(typeof(T), out var lazyBindDescriptor))
+            {
+                var instance = (T) InstantiateLazyBoundObject(lazyBindDescriptor);
+                m_LazyBindDescriptors.Remove(typeof(T));
+                return instance;
+            }
+
+            if (!m_TypesToInstances.TryGetValue(typeof(T), out var value))
+            {
+                if (m_Parent != null) return m_Parent.Resolve<T>();
+
+                throw new NoInstanceToInjectException($"Injection of type {typeof(T)} failed.");
+            }
+
+            return (T) value;
+        }
+
+        public void InjectIn(object obj)
+        {
+            if (CachedReflectionUtility.TryGetInjectableMethod(obj.GetType(), out var injectionMethod))
+            {
+                var parameters = CachedReflectionUtility.GetMethodParameters(injectionMethod);
+
+                var paramColleciton = new object[parameters.Length];
+
+                for (var i = 0; i < parameters.Length; i++)
+                {
+                    var parameter = parameters[i];
+
+                    var genericResolveMethod = CachedReflectionUtility.GetTypedResolveMethod(parameter.ParameterType);
+                    var resolved = genericResolveMethod.Invoke(this, null);
+                    paramColleciton[i] = resolved;
+                }
+
+                injectionMethod.Invoke(obj, paramColleciton);
+            }
+        }
+
+        public void InjectIn(GameObject go)
+        {
+            var components = go.GetComponentsInChildren<Component>();
+
+            foreach (var component in components) InjectIn(component);
+        }
+
+        ~DIScope()
+        {
+            Dispose();
+        }
+
+        public void BindInstanceAsSingle<T>(T instance) where T : class
+        {
+            BindInstanceToType(instance, typeof(T));
+        }
+
+        public void BindInstanceAsSingle<TInterface, TImplementation>(TImplementation instance)
+            where TImplementation : class, TInterface
+            where TInterface : class
+        {
+            BindInstanceAsSingle<TInterface>(instance);
+            BindInstanceAsSingle(instance);
+        }
+
+        private void BindInstanceToType(object instance, Type type)
+        {
+            m_TypesToInstances[type] = instance;
+        }
+
+        public void BindAsSingle<TImplementation, TInterface>()
```

Comment:
Any reason this has the same name as BindAsSingle with only one type?
Basically, this is a way to bind a default implementation for a given interface correct? Just "BindAsSingle" might be a bit confusing for that.
On another hand we don't want BindDefaultImplementationToInterfaceAsSingle...

Reply:
Yeah, there is no specific reason to go with an overloaded function, however I just have a preference for overloads when they seem to make sense. So the intent here is to allow the user to bind a concrete implementation type that will be available by it's concrete type (that's a default of the DIScope - if something is bound, it is always automatically bound to (at the very least) it's concrete type) AND by the interface type.  

I'll try and see if there are some better groupings and names for these method - lmk if you have ideas for short and concise intentful namings.

Reply:
BindInterfaceImplementationAsSingle seems like a mouthful too.
BindInterfaceAsSingle?